sys/net/gnrc/routing/ipv6_auto_subnets/gnrc_ipv6_auto_subnets.c

```diff
@@ -0,0 +1,221 @@
+        if (match_len > old_pfx_len) {
+            old_pfx_len = entry.pfx_len;
+            old_pfx = entry.pfx;
+        }
+    }
+
+    /* no prefix found */
+    if (old_pfx_len == 0) {
+        return false;
+    }
+
+    DEBUG("auto_subnets: remove old prefix %s/%u\n",
+          ipv6_addr_to_str(addr_str, &old_pfx, sizeof(addr_str)), old_pfx_len);
+
+    /* invalidate old prefix in RIO */
+    tmp = gnrc_ndp_opt_ri_build(&old_pfx, old_pfx_len, 0,
+                                NDP_OPT_RI_FLAGS_PRF_NONE, *ext_opts);
+    if (tmp) {
+        *ext_opts = tmp;
+    }
+
+    /* remove the prefix */
+    gnrc_ipv6_nib_pl_del(netif->pid, &old_pfx, old_pfx_len);
+
+    return false;
+}
+
+static void _configure_subnets(uint8_t subnets, gnrc_netif_t *upstream,
+                               const ndp_opt_pi_t *pio)
+{
+    gnrc_netif_t *downstream = NULL;
+    gnrc_pktsnip_t *ext_opts = NULL;
+    const ipv6_addr_t *prefix = &pio->prefix;
+    uint32_t valid_ltime = byteorder_ntohl(pio->valid_ltime);
+    uint32_t pref_ltime = byteorder_ntohl(pio->pref_ltime);
+    const uint8_t prefix_len = pio->prefix_len;
+    uint8_t new_prefix_len, subnet_len;
+
+    DEBUG("auto_subnets: create %u subnets\n", subnets);
+
+    /* Calculate remaining prefix length. For n subnets we consume ⌊log₂ n⌋ + 1 bits.
+     * To calculate ⌊log₂ n⌋ quickly, find the position of the most significant set bit
+     * by counting leading zeros.
+     */
+    subnet_len = 32 - __builtin_clz(subnets);
+    new_prefix_len = prefix_len + subnet_len;
+
+    if (new_prefix_len > 64) {
+        DEBUG("auto_subnets: can't split /%u into %u subnets\n", prefix_len, subnets);
+        return;
+    }
+
+    while ((downstream = gnrc_netif_iter(downstream))) {
+        gnrc_pktsnip_t *tmp;
+        ipv6_addr_t new_prefix;
+
+        if (downstream == upstream) {
+            continue;
+        }
+
+        /* create subnet from upstream prefix */
+        _init_sub_prefix(&new_prefix, prefix, prefix_len, subnets--, subnet_len);
+
+        DEBUG("auto_subnets: configure prefix %s/%u on %u\n",
+              ipv6_addr_to_str(addr_str, &new_prefix, sizeof(addr_str)),
+              new_prefix_len, downstream->pid);
+
+        /* first remove old prefix if the prefix changed */
+        _remove_old_prefix(downstream, &new_prefix, new_prefix_len, &ext_opts);
+
+        /* configure subnet on downstream interface */
+        gnrc_netif_ipv6_add_prefix(downstream, &new_prefix, new_prefix_len,
+                                   valid_ltime, pref_ltime);
+
+        /* start advertising subnet */
+        gnrc_ipv6_nib_change_rtr_adv_iface(downstream, true);
+
+        /* add route information option with new subnet */
+        tmp = gnrc_ndp_opt_ri_build(&new_prefix, new_prefix_len, valid_ltime,
+                                    NDP_OPT_RI_FLAGS_PRF_NONE, ext_opts);
+        if (tmp == NULL) {
+            DEBUG("auto_subnets: No space left in packet buffer. Not adding RIO\n");
```

Comment:
I might be mistaken, but I think, in case `ext_opts` was already set (as this is a loop this well may be and also with `_remove_old_prefix()`), you need to release `ext_opts` in this case. Otherwise there is a potential packet buffer leak here.

Reply:
That's why I used the `tmp` variable. `ext_opts` is only overwritten if `tmp` is not NULL, then `gnrc_ndp_rtr_adv_send()` is going to take care of it.

Reply:
Yeah, but in previous iterations or in `_remove_old_prefix()` `ext_opts` might have been allocated which is free'd nowhere now, right?

Reply:
Ah,... you are not bailing here! But maybe add a break instead of trying to fill up an already filled packet buffer?

Reply:
Ping?

Reply:
I thought it would be still worthwhile to configure the downstream networks even if the packet buffer is currently exhausted to inform upstream.

We might also bail out here to make the breakage more visible.

Reply:
> I thought it would be still worthwhile to configure the downstream networks even if the packet buffer is currently exhausted to inform upstream.
> 
> We might also bail out here to make the breakage more visible.

Might make sense. And there is a chance, that some other module gives up resources again, so maybe we do not need to `break;` here. Ok!